### PR TITLE
Various verification fixes

### PR DESCRIPTION
**Pull request by gnardini**
### Description

Changes in this PR (some are further explained with a PR comment in the relevant place):
- Only run SMS retrieval on Android
- If user was already verified, skip verification (it was getting stuck in the loading screen).
- If user runs out of Komenci Quota, go back to `VerificationEducationScreen` to redo recaptcha and get more.
- When resending messages, resend only the non completed ones and cancel the previous code listeners that were waiting for codes to be input, otherwise they are processed twice and one of the tries fails, which cancels the whole verification flow.
- Fix an issue with locking: We were not acquiring the lock correctly, so we got some errors due to race conditions sometimes. See PR comment for details.
- Implement lock in `receiveAttestationCodeSaga` to avoid UI glitch if many codes are paster **very** quickly together.
- Catch all sagas with an error handler that fires the fail saga if verification fails (it was getting stuck instead of showing the error modal in those cases).

### Other changes

N/A

### Tested

Manually many times. Some edge cases I tested: 
- I put some sleeps in the code to test the locks.
- I wrote code to input the three codes at the same time.
- I requested a bunch of attestations to run out of quota.
- I resent messages a few times within the same session and across different app runs.
- I input the codes in random order in every test.

### How others should test

The whole verification flow needs to be tested including all edge cases that one can think of.

### Related issues

- Fixes #377

### Backwards compatibility

N/A

### CODE REVIEW

File: packages/mobile/src/verify/komenci.ts [hunk 1]
```diff
@@ -143,8 +143,7 @@ export function* fetchOrDeployMtwSaga() {
     // user already has a verified MTW
     const verifiedMtwAddress = yield call(fetchVerifiedMtw, contractKit, walletAddress)
     if (verifiedMtwAddress) {
-      // TODO: Address this
-      // yield put(doVerificationFlow(true))
+      yield put(succeed())
```
Comment (gnardini) [inline packages/mobile/src/verify/komenci.ts:146]: Not succeeding here was causing already verified users to get stuck in the verification flow with no actions to do.

File: packages/mobile/src/verify/saga.ts [hunk 2]
```diff
@@ -197,19 +204,24 @@ function* getCodeForIssuer(issuer: string): Generator<any, AttestationCode | und
...
     return code
   }
 
-  while (true) {
-    const action: ReturnType<typeof inputAttestationCode> = yield take(inputAttestationCode.type)
-    if (action.payload.issuer === issuer) {
-      return action.payload
-    }
-  }
+  const {
```
Comment (gnardini) [inline packages/mobile/src/verify/saga.ts:214]: Not canceling here meant that when a user re-requested codes there were two listeners per code, the original one and the one created for the resend, so when the user put in the code for the issuer 2 codes were processed and one failed, causing the user to fail out of the flow.

File: packages/mobile/src/verify/saga.ts [hunk 3]
```diff
@@ -291,12 +306,8 @@ export function* completeAttestation(
     `Completing code (${code.shortCode} ${codePosition}) for issuer: ${code.issuer}`
   )
 
-  // Make each concurrent completion attempt wait a sec for where they are relative to other codes
-  // to ensure `processingInputCode` has enough time to properly gate the tx. 0-index code
-  // will have 0 delay, 1-index code will have 1 sec delay, etc.
   if (shouldUseKomenci) {
-    yield delay(codePosition * 5000)
-    yield inputAttestationCodeLock.acquireAsync()
+    yield call([inputAttestationCodeLock, inputAttestationCodeLock.acquireAsync])
```
Comment (gnardini) [inline packages/mobile/src/verify/saga.ts:310]: The delay was necessary because the lock wasn't working as intended because it was not being called correctly. redux-saga can be non-intuitive at times :/

File: packages/mobile/src/verify/saga.ts [hunk 4]
```diff
@@ -603,35 +612,34 @@ export function* completeAttestationsSaga() {
...
+  Logger.debug(TAG, `@resendMessagesSaga has started`)
+
   const shouldUseKomenci = yield select(shouldUseKomenciSelector)
   const status = yield select(verificationStatusSelector)
   ValoraAnalytics.track(VerificationEvents.verification_resend_messages, {
     count: status.numAttestationsRemaining,
     feeless: shouldUseKomenci,
   })
 
-  const revealStatuses = yield select(revealStatusesSelector)
```
Comment (gnardini) [inline packages/mobile/src/verify/saga.ts:619]: I'm not sure what was the thinking behind this code, but it wasn't working. Hopefully it's more intuitive with the changes, we just set non confirmed issuers as `NotReveled` and call `revealAttestations `

File: packages/mobile/src/verify/saga.ts [hunk 5]
```diff
@@ -685,34 +693,38 @@ export function* receiveAttestationCodeSaga(action: ReturnType<typeof receiveAtt
...
-    Logger.error(
-      TAG + '@attestationCodeReceiver',
-      'All attestation code positions are full. Ignoring.'
-    )
-    return
-  }
-  yield put(setAttestationInputStatus({ index, status: CodeInputStatus.Received }))
+  // Only one code should be being received at the same time so that it doesn't cause any UI glitches by using
+  // the same index position for more that one. We use the lock for that.
+  yield call([receiveAttestationCodeLock, receiveAttestationCodeLock.acquireAsync])
```
Comment (gnardini) [inline packages/mobile/src/verify/saga.ts:698]: The diff here is a bit hard to read but all I did is move the `try` of the `try/catch` earlier in the code and put a lock (`receiveAttestationCodeLock`) so that only one code goes through this part at the time.

File: packages/mobile/src/verify/saga.ts [hunk 6]
```diff
@@ -795,6 +807,7 @@ export function* receiveAttestationCodeSaga(action: ReturnType<typeof receiveAtt
       `Attestation code (${message}) is valid, starting processing (issuer: ${issuer})`
     )
 
+    yield put(setAttestationInputStatus({ index, status: CodeInputStatus.Processing }))
```
Comment (gnardini) [inline packages/mobile/src/verify/saga.ts:810]: I lied in the previous comment, I also added this which is necessary so that users can input more codes while the first one completes.

File: packages/mobile/src/verify/saga.ts [hunk 7]
```diff
@@ -822,35 +839,64 @@ export function* getPhoneHashDetails() {
...
+      yield call(saga, ...args)
+    } catch (error) {
+      let errorString = error
+      let lastError = error
+      if (Array.isArray(error)) {
+        errorString = error.join(' - ')
+        lastError = error[-1]
+      }
+      Logger.error(TAG, `Error while running saga ${sagaName}`, error)
+      if (lastError?.toString() === FetchErrorTypes.QuotaExceededError) {
```
Comment (gnardini) [inline packages/mobile/src/verify/saga.ts:854]: This happens if the user runs out of Komenci quota. Can happen if they re-request attestations too many times.

File: packages/mobile/src/verify/saga.ts [hunk 8]
```diff
@@ -822,35 +845,64 @@ export function* getPhoneHashDetails() {
...
+        yield put(resetKomenciSession())
+      } else {
+        yield put(fail(errorString))
```
Comment (jeanregisser) [inline packages/mobile/src/verify/saga.ts:863]: Does this display a message to the user?

**Review by jeanregisser — COMMENTED** [review]
This is great! 👍 

Nothing to do now, but this shows we don't have enough unit test coverage around this critical part.
Hopefully we can address this at some point 👀

File: packages/mobile/src/verify/saga.ts [hunk 9]
```diff
@@ -805,7 +824,11 @@ export function* receiveAttestationCodeSaga(action: ReturnType<typeof receiveAtt
       error
     )
     yield put(showError(ErrorMessages.INVALID_ATTESTATION_CODE))
-    yield put(setAttestationInputStatus({ index, status: CodeInputStatus.Error }))
+    if (index) {
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:827]: Yes, thanks

File: packages/mobile/src/verify/sagas/resetKomenciSessionSaga.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { navigate } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+
+// Should this not be part of the saga? When we find out a user ran out of Komenci quota we need
+// them to redo the reCaptcha. To do that they need to go back to the VerificationEducationScreen.
+// When this happens they are usually on the VerificationInputScreen, so this brings them back so
+// they have to press on 'start' again and they get the reCaptcha prompt.
+export default function resetKomenciSessionSaga() {
+  navigate(Screens.VerificationEducationScreen)
+}
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/verify/sagas/resetKomenciSessionSaga.ts:10]: The main `saga` file has 900+ lines, so I didn't want to put more stuff there. I would like to move all sagas each to their different file. Ivan had started doing this but only did it for a few, it'd be ideal to have each saga on its own file which would tidy up the folder structure imo.

Since I haven't moved the others yet I could move this to the main saga file for now if you prefer

File: packages/mobile/src/verify/saga.ts (hunk 8 above)
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:863]: Yes, calling `fail` sets the verification state to `Failed` and shows the 'Confirmation Failed' modal with a 'Skip for now' CTA

File: packages/mobile/src/verify/saga.ts [hunk 11]
```diff
@@ -822,35 +845,64 @@ export function* getPhoneHashDetails() {
...
+      if (lastError?.toString() === FetchErrorTypes.QuotaExceededError) {
+        yield put(resetKomenciSession())
+      } else {
+        yield put(fail(errorString))
+      }
+    }
+  }
+}
+
+const sagas: Array<[string, any]> = [
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:869]: Yeah, I spent some time on this but couldn't figure out the typing. I'm sure there's a way to make it work, but I didn't want to spend too much more time on this :/

File: packages/mobile/src/verify/saga.ts (hunk 8 above)
Comment (jeanregisser) [reply packages/mobile/src/verify/saga.ts:863]: Good, and it's always a user friendly error, right? not the error string we here?

**Review by jeanregisser — APPROVED** [review]

File: packages/mobile/src/verify/saga.ts (hunk 8 above)
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:863]: Yes, the error string is sent to analytics, the text of the modal is generic and fixed

File: packages/mobile/src/verify/requestAttestations.ts [hunk 12]
```diff
@@ -113,12 +112,9 @@ export function* requestAttestationsSaga() {
       feeless: shouldUseKomenci,
     })
     yield put(revealAttestations())
-  } catch (e) {
-    if (shouldUseKomenci) {
-      yield put(startKomenciSession())
-    } else {
-      yield put(fail(e.message))
-    }
+  } catch (error) {
```
Comment (silasbw) [inline packages/mobile/src/verify/requestAttestations.ts:115]: nit: remove the try/catch here?  or is there a specific reason to have the `catch (error) { throw error }` pattern?

**Review by silasbw — COMMENTED** [review]
if this fixed a bunch of issues, should we have tests (e.g., end-to-end) that assert the correct behavior?

File: packages/mobile/src/verify/saga.ts (hunk 3 above)
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:312]: I'll update the comment and add a TODO in the code to revisit that

File: packages/mobile/src/verify/saga.ts (hunk 6 above)
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:816]: When codes are input there are two steps: 
- The first one is this function which does some basic and quick checks. 
- The next one is the complete saga, which calls komenci to send the codes.

When a user puts in a code, we want them to wait until we do the first few checks before going to the next input. Otherwise, if these checks fail the user will already have the next input focused but the previous one will also be available because it failed. This is captured by the `CodeInputStatus`, `Received` means doing some quick checks and `Processing` means sending the codes to Komenci.

There might be a better way of doing this, maybe we can just make this quick checks in the UI so that the `Received` status is no longer necessary (we'd have to check automated code-pasting from android too) but the UI still depends on these index-based mappings unfortunately :(

File: packages/mobile/src/verify/saga.ts [hunk 13]
```diff
@@ -603,35 +617,35 @@ export function* completeAttestationsSaga() {
     })
   )
 
-  yield put(setMtwAddress(komenci.unverifiedMtwAddress))
-  yield put(succeed())
+  // If some code was canceled we haven't actually finished :)
+  if (results.every((result) => result)) {
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:621]: I don't think that can happen but you're suggestion is probably safer either way, made the change :)

File: packages/mobile/src/verify/requestAttestations.ts (hunk 12 above)
Comment (gnardini) [reply packages/mobile/src/verify/requestAttestations.ts]: No reason, removed

**Comment by gnardini** [comment]
> if this fixed a bunch of issues, should we have tests (e.g., end-to-end) that assert the correct behavior?
Completely agree we need tests for this flow, both unit and e2e. I think it's telling that I did all these changes and not a single test broke.
I didn't add them here because there's a time constraint of having this ready before the 1.15 branch cut and I think @etuleu and @MuckT are already working on adding some, but I'll give them a hand to improve the testing of this flow which is sorely lacking tests.

**Comment by gnardini** [comment]
@tarikbellamine 
> However, the fact that mistakes are still so easy to make and hard to spot is worrisome to me. How can we improve the structure of the flow to improve readability and cut down on errors? IMO the lack of typing on redux-sagas is a BIG contributor to some of these issues. Can you think of anything else?

I agree with and share the worry. The `verify/saga` file is 900+ lines long, something easy to do is to separate all the sagas to their own file. That will help visualize each step in isolation. I also think that the current redux-based communication between different sagas is not ideal. I thought the refactor was meant to turn this into more of a state-machine but that doesn't seem to be the case.

Basically, right now each saga is responsible to either call another one or wait for the previous one. Instead of this, I'd love for each saga to only do what it's responsible for and have an orchestrator of sorts that calls the sagas and decides what to do with the response - either throw an error or call another one. This will allow us to separate the decision-making of _what_ to do from the implementation details of _how_ it's done. A possibility for doing this is just have the orchestrator be a saga and the rest be simple async functions. 

One property of sagas is that each individual one can access all the redux state and fire actions, but that might make things complicated since there are side-effects being fired from all over the place instead of having everything focused in a single place. Might be worth an exploration to move away from sagas.

File: packages/mobile/src/verify/saga.ts (hunk 6 above)
Comment (tarikbellamine) [reply packages/mobile/src/verify/saga.ts:821]: I see. Is there still value in restricting inputs though? Why do we even focus code inputs one by one instead of just making them all available at the same time? Maybe this was to make sure that codes were inputting sequentially but now that we have mutex locks it may just be adding unnecessary complexity

**Comment by tarikbellamine** [comment]
> One property of sagas is that each individual one can access all the redux state and fire actions, but that might make things complicated since there are side-effects being fired from all over the place instead of having everything focused in a single place. Might be worth an exploration to move away from sagas.

I think I agree with this. It's super complicated how all the different retry logic and error throwing/handling is done within every saga and there are a large amount of side effects that make it impossible to follow the code path of a given user. I hadn't considered getting rid of sagas altogether but that does seem to both reduce side effects and help introduce more typing.

**Comment by Celoqa** [comment]
Hi @tarikbellamine @i1skn We have verified this on latest Android internal build V1.15.0 & Testflight build V1.15.0(63) and observe the followings:

**Verification flow with create new account:**
- [x] User is able to complete the verification flow without any issues while auto pasting the codes. 
- [x] User is able to Resend the codes & complete the verification flow without any gaps.

**Verification flow with Restore account key:**

- [x] User is able to complete the verification flow without any issues while auto pasting the codes. 
- [x] User is able to Resend the codes & complete the verification flow without any gaps.

**Restore account with already verified number:**

- [x] User directly redirects to home Connect contacts page when try to restore account with already verified number

Also we have performed ["Phone number verification"](https://valora.testquality.com/project/12128/plan/20403/run/148219/result/2412149) cases with different scenarios on Testquality tool.

**Comment by MuckT** [comment]
https://valora.testquality.com requirement [R34](https://valora.testquality.com/project/12128/testcase/286105/requirement/11367). Do not remove or edit.

File: packages/mobile/src/verify/saga.ts (hunk 6 above)
Comment (gnardini) [reply packages/mobile/src/verify/saga.ts:825]: Oh, I had missed this comment. I'm not sure, maybe a UX decision to give users clarity on what they have to do? If they have the 3 options they may not know where to put the code or something. Definitely worth thinking about this

**Comment by Celoqa** [comment]
Hi @tarikbellamine we have verified phone number confirmation in android internal build V1.15.2 and observed the following.
- Second code is shown blank  after auto pasting in android 
However user is able to go further without second code.
[Attachment](https://drive.google.com/file/d/1rGwV0_TdzDrDV7m6zpVXSMqBnwv00brc/view?usp=sharing)
